File: app/src/main/java/com/stevenfrew/beatprompter/cache/CachedCloudFileCollection.java
package com.stevenfrew.beatprompter.cache;

import android.util.Log;

import com.stevenfrew.beatprompter.BeatPrompterApplication;
import com.stevenfrew.beatprompter.cloud.CloudFileInfo;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.w3c.dom.*;

public class CachedCloudFileCollection
{
    private List<CachedCloudFile> mFiles=new ArrayList<>();

    public void writeToXML(Document d,Element root)
    {
        for(CachedCloudFile ccf:mFiles)
            ccf.writeToXML(d,root);
    }

    public void readFromXML(Document xmlDoc)
    {
        clear();

        NodeList songFiles = xmlDoc.getElementsByTagName(SongFile.SONGFILE_ELEMENT_TAG_NAME);
        for (int f = 0; f < songFiles.getLength(); ++f) {
            Node n = songFiles.item(f);
            SongFile song = new SongFile((Element)n);
            add(song);
        }
        NodeList setFiles = xmlDoc.getElementsByTagName(SetListFile.SETLISTFILE_ELEMENT_TAG_NAME);
        for (int f = 0; f < setFiles.getLength(); ++f) {
            Node n = setFiles.item(f);
            try {
                SetListFile set = new SetListFile((Element) n);
                add(set);
            }
            catch(InvalidBeatPrompterFileException ibpfe)
            {
                // This should never happen. If we could write out the file info, then it was valid.
                // So it must still be valid when we come to read it in. Unless some dastardly devious sort
                // has meddled with files outside of the app ...
                Log.d(BeatPrompterApplication.TAG,"Failed to parse set-list file.");
            }
        }
        NodeList imageFiles = xmlDoc.getElementsByTagName(ImageFile.IMAGEFILE_ELEMENT_TAG_NAME);
        for (int f = 0; f < imageFiles.getLength(); ++f) {
            Node n = imageFiles.item(f);
            ImageFile imageFile = new ImageFile((Element)n);
            add(imageFile);
        }
        NodeList audioFiles = xmlDoc.getElementsByTagName(AudioFile.AUDIOFILE_ELEMENT_TAG_NAME);
        for (int f = 0; f < audioFiles.getLength(); ++f) {
            Node n = audioFiles.item(f);
            AudioFile audioFile = new AudioFile((Element)n);
            add(audioFile);
        }
        NodeList aliasFiles = xmlDoc.getElementsByTagName(MIDIAliasFile.MIDIALIASFILE_ELEMENT_TAG_NAME);
        for (int f = 0; f < aliasFiles.getLength(); ++f) {
            Node n = aliasFiles.item(f);
            try{
                MIDIAliasFile midiAliasCachedCloudFile = new MIDIAliasFile((Element)n);
                add(midiAliasCachedCloudFile);
            }
            catch(InvalidBeatPrompterFileException ibpfe)
            {
                // This should never happen. If we could write out the file info, then it was valid.
                // So it must still be valid when we come to read it in. Unless some dastardly devious sort
                // has meddled with files outside of the app ...
                Log.d(BeatPrompterApplication.TAG,"Failed to parse MIDI alias file.");
            }
        }
    }

    public void add(CachedCloudFile cachedFile)
    {
        for(int f=mFiles.size()-1;f>=0;--f)
        {
            CachedCloudFile existingFile=mFiles.get(f);
            if(cachedFile.mID.equalsIgnoreCase(existingFile.mID)) {
                mFiles.set(f, cachedFile);
                return;
            }
        }
        mFiles.add(cachedFile);
    }

    public void remove(CloudFileInfo cloudFile)
    {
        for(int f=mFiles.size()-1;f>=0;--f)
        {
            CachedCloudFile existingFile=mFiles.get(f);
            if(cloudFile.mID.equalsIgnoreCase(existingFile.mID))
                mFiles.remove(f);
        }
    }

    public boolean hasLatestVersionOf(CloudFileInfo cloudFile)
    {
        return mFiles.stream().anyMatch(f->f.mID.equalsIgnoreCase(cloudFile.mID) && f.mLastModified.equals(cloudFile.mLastModified));
    }

    public List<SongFile> getSongFiles()
    {
        return mFiles.stream().filter(f->f instanceof SongFile).map(f->(SongFile)f).collect(Collectors.toList());
    }

    public List<SetListFile> getSetListFiles()
    {
        return mFiles.stream().filter(f->f instanceof SetListFile).map(f->(SetListFile)f).collect(Collectors.toList());
    }

    public List<MIDIAliasFile> getMIDIAliasFiles()
    {
        return mFiles.stream().filter(f->f instanceof MIDIAliasFile).map(f->(MIDIAliasFile)f).collect(Collectors.toList());
    }

    private List<AudioFile> getAudioFiles()
    {
        return mFiles.stream().filter(f->f instanceof AudioFile).map(f->(AudioFile)f).collect(Collectors.toList());
    }

    private List<ImageFile> getImageFiles()
    {
        return mFiles.stream().filter(f->f instanceof ImageFile).map(f->(ImageFile)f).collect(Collectors.toList());
    }

    public void removeNonExistent(Set<String> storageIDs)
    {
        List<CachedCloudFile> remainingFiles=mFiles.stream().filter(f->storageIDs.contains(f.mID)).collect(Collectors.toList());
        List<CachedCloudFile> noLongerExistingFiles=mFiles.stream().filter(f->!storageIDs.contains(f.mID)).collect(Collectors.toList());
        noLongerExistingFiles.forEach(f->{if(!f.mFile.delete())
            Log.e(BeatPrompterApplication.TAG, "Failed to delete file: "+f.mFile.getName());});
        mFiles=remainingFiles;
    }

    public void clear()
    {
        mFiles.clear();
    }

    public boolean isEmpty()
    {
        return mFiles.isEmpty();
    }

    public AudioFile getMappedAudioFilename(String in,ArrayList<AudioFile> tempAudioFileCollection)
    {
        if(in!=null) {
            for (AudioFile afm : getAudioFiles()) {
                String secondChance = in.replace('’', '\'');
                if ((afm.mName.equalsIgnoreCase(in)) || (afm.mName.equalsIgnoreCase(secondChance)))
                    return afm;
            }
            if(tempAudioFileCollection!=null)
                for (AudioFile afm : tempAudioFileCollection) {
                    String secondChance = in.replace('’', '\'');
                    if ((afm.mName.equalsIgnoreCase(in)) || (afm.mName.equalsIgnoreCase(secondChance)))
                        return afm;
                }
        }
        return null;
    }

    public ImageFile getMappedImageFilename(String in,ArrayList<ImageFile> tempImageFileCollection)
    {
        if(in!=null) {
            for (ImageFile ifm : getImageFiles()) {
                String secondChance = in.replace('’', '\'');
                if ((ifm.mName.equalsIgnoreCase(in)) || (ifm.mName.equalsIgnoreCase(secondChance)))
                    return ifm;
            }
            if(tempImageFileCollection!=null)
                for (ImageFile ifm : tempImageFileCollection) {
                    String secondChance = in.replace('’', '\'');
                    if ((ifm.mName.equalsIgnoreCase(in)) || (ifm.mName.equalsIgnoreCase(secondChance)))
                        return ifm;
                }
        }
        return null;
    }

    public ArrayList<CachedCloudFile> getFilesToRefresh(CachedCloudFile fileToRefresh, boolean includeDependencies)
    {
        ArrayList<CachedCloudFile> filesToRefresh=new ArrayList<>();
        if(fileToRefresh!=null)
        {
            filesToRefresh.add(fileToRefresh);
            if((fileToRefresh instanceof SongFile) && (includeDependencies))
            {
                SongFile song=(SongFile)fileToRefresh;
                if (song.mAudioFiles != null)
                    for (String audioFileName : song.mAudioFiles) {
                        AudioFile audioFile = getMappedAudioFilename(audioFileName, null);
                        File actualAudioFile = null;
                        if (audioFile != null)
                            actualAudioFile = new File(song.mFile.getParent(), audioFile.mFile.getName());
                        if ((actualAudioFile != null) && (actualAudioFile.exists()))
                            filesToRefresh.add(audioFile);
                    }
                if (song.mImageFiles != null)
                    for (String imageFileName : song.mImageFiles) {
                        ImageFile imageFile = getMappedImageFilename(imageFileName, null);
                        File actualImageFile = null;
                        if (imageFile != null)
                            actualImageFile = new File(song.mFile.getParent(), imageFile.mFile.getName());
                        if ((actualImageFile != null) && (actualImageFile.exists()))
                            filesToRefresh.add(imageFile);
                    }
            }
        }
        return filesToRefresh;
    }


}

File: app/src/main/java/com/stevenfrew/beatprompter/cache/CachedCloudFileCollection.kt
package com.stevenfrew.beatprompter.cache

import android.util.Log
import com.stevenfrew.beatprompter.BeatPrompterApplication
import com.stevenfrew.beatprompter.cloud.CloudFileInfo
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.io.File
import java.util.ArrayList

class CachedCloudFileCollection {
    private var mFiles: MutableList<CachedCloudFile> = ArrayList()

    val songFiles: List<SongFile>
        get() =
            mFiles.filterIsInstance<SongFile>()

    val setListFiles: List<SetListFile>
        get() =
            mFiles.filterIsInstance<SetListFile>()

    val midiAliasFiles: List<MIDIAliasFile>
        get() =
            mFiles.filterIsInstance<MIDIAliasFile>()

    private val audioFiles: List<AudioFile>
        get() =
            mFiles.filterIsInstance<AudioFile>()

    private val imageFiles: List<ImageFile>
        get() =
            mFiles.filterIsInstance<ImageFile>()

    val isEmpty: Boolean
        get() = mFiles.isEmpty()

    fun writeToXML(d: Document, root: Element) {
        for (ccf in mFiles)
            ccf.writeToXML(d, root)
    }

    fun readFromXML(xmlDoc: Document) {
        clear()

        val songFiles = xmlDoc.getElementsByTagName(SongFile.SONGFILE_ELEMENT_TAG_NAME)
        for (f in 0 until songFiles.length) {
            val n = songFiles.item(f)
            val song = SongFile(n as Element)
            add(song)
        }
        val setFiles = xmlDoc.getElementsByTagName(SetListFile.SETLISTFILE_ELEMENT_TAG_NAME)
        for (f in 0 until setFiles.length) {
            val n = setFiles.item(f)
            try {
                val set = SetListFile(n as Element)
                add(set)
            } catch (ibpfe: InvalidBeatPrompterFileException) {
                // This should never happen. If we could write out the file info, then it was valid.
                // So it must still be valid when we come to read it in. Unless some dastardly devious sort
                // has meddled with files outside of the app ...
                Log.d(BeatPrompterApplication.TAG, "Failed to parse set-list file.")
            }

        }
        val imageFiles = xmlDoc.getElementsByTagName(ImageFile.IMAGEFILE_ELEMENT_TAG_NAME)
        for (f in 0 until imageFiles.length) {
            val n = imageFiles.item(f)
            val imageFile = ImageFile(n as Element)
            add(imageFile)
        }
        val audioFiles = xmlDoc.getElementsByTagName(AudioFile.AUDIOFILE_ELEMENT_TAG_NAME)
        for (f in 0 until audioFiles.length) {
            val n = audioFiles.item(f)
            val audioFile = AudioFile(n as Element)
            add(audioFile)
        }
        val aliasFiles = xmlDoc.getElementsByTagName(MIDIAliasFile.MIDIALIASFILE_ELEMENT_TAG_NAME)
        for (f in 0 until aliasFiles.length) {
            val n = aliasFiles.item(f)
            try {
                val midiAliasCachedCloudFile = MIDIAliasFile(n as Element)
                add(midiAliasCachedCloudFile)
            } catch (ibpfe: InvalidBeatPrompterFileException) {
                // This should never happen. If we could write out the file info, then it was valid.
                // So it must still be valid when we come to read it in. Unless some dastardly devious sort
                // has meddled with files outside of the app ...
                Log.d(BeatPrompterApplication.TAG, "Failed to parse MIDI alias file.")
            }

        }
    }

    fun add(cachedFile: CachedCloudFile) {
        for (f in mFiles.indices.reversed()) {
            val existingFile = mFiles[f]
            if (cachedFile.mID.equals(existingFile.mID, ignoreCase = true)) {
                mFiles[f] = cachedFile
                return
            }
        }
        mFiles.add(cachedFile)
    }

    fun remove(cloudFile: CloudFileInfo) {
        for (f in mFiles.indices.reversed()) {
            val existingFile = mFiles[f]
            if (cloudFile.mID.equals(existingFile.mID, ignoreCase = true))
                mFiles.removeAt(f)
        }
    }

    fun hasLatestVersionOf(cloudFile: CloudFileInfo): Boolean {
        return mFiles.stream().anyMatch { f -> f.mID.equals(cloudFile.mID, ignoreCase = true) && f.mLastModified == cloudFile.mLastModified }
    }

    fun removeNonExistent(storageIDs: Set<String>) {
        val remainingFiles=mFiles.filter{storageIDs.contains(it.mID)}
        val noLongerExistingFiles=mFiles.filter{!storageIDs.contains(it.mID)}
//        val remainingFiles = mFiles.stream().filter { f -> storageIDs.contains(f.mID) }.collect<List<CachedCloudFile>, Any>(Collectors.toList())
//        val noLongerExistingFiles = mFiles.stream().filter { f -> !storageIDs.contains(f.mID) }.collect<List<CachedCloudFile>, Any>(Collectors.toList())
        noLongerExistingFiles.forEach { f ->
            if (!f.mFile.delete())
                Log.e(BeatPrompterApplication.TAG, "Failed to delete file: " + f.mFile.name)
        }
        mFiles = remainingFiles.toMutableList()
    }

    fun clear() {
        mFiles.clear()
    }

    fun getMappedAudioFilename(`in`: String?, tempAudioFileCollection: ArrayList<AudioFile>?): AudioFile? {
        if (`in` != null) {
            for (afm in audioFiles) {
                val secondChance = `in`.replace('’', '\'')
                if (afm.mName.equals(`in`, ignoreCase = true) || afm.mName.equals(secondChance, ignoreCase = true))
                    return afm
            }
            if (tempAudioFileCollection != null)
                for (afm in tempAudioFileCollection) {
                    val secondChance = `in`.replace('’', '\'')
                    if (afm.mName.equals(`in`, ignoreCase = true) || afm.mName.equals(secondChance, ignoreCase = true))
                        return afm
                }
        }
        return null
    }

    fun getMappedImageFilename(`in`: String?, tempImageFileCollection: ArrayList<ImageFile>?): ImageFile? {
        if (`in` != null) {
            for (ifm in imageFiles) {
                val secondChance = `in`.replace('’', '\'')
                if (ifm.mName.equals(`in`, ignoreCase = true) || ifm.mName.equals(secondChance, ignoreCase = true))
                    return ifm
            }
            if (tempImageFileCollection != null)
                for (ifm in tempImageFileCollection) {
                    val secondChance = `in`.replace('’', '\'')
                    if (ifm.mName.equals(`in`, ignoreCase = true) || ifm.mName.equals(secondChance, ignoreCase = true))
                        return ifm
                }
        }
        return null
    }

    fun getFilesToRefresh(fileToRefresh: CachedCloudFile?, includeDependencies: Boolean): ArrayList<CachedCloudFile> {
        val filesToRefresh = ArrayList<CachedCloudFile>()
        if (fileToRefresh != null) {
            filesToRefresh.add(fileToRefresh)
            if (fileToRefresh is SongFile && includeDependencies) {
                val song = fileToRefresh as SongFile?
                if(song!=null) {
                    for (audioFileName in song.mAudioFiles) {
                        val audioFile = getMappedAudioFilename(audioFileName, null)
                        var actualAudioFile: File? = null
                        if (audioFile != null)
                            actualAudioFile = File(song.mFile.parent, audioFile.mFile.name)
                        if (actualAudioFile != null && actualAudioFile.exists())
                            filesToRefresh.add(audioFile!!)
                    }
                    for (imageFileName in song.mImageFiles) {
                        val imageFile = getMappedImageFilename(imageFileName, null)
                        var actualImageFile: File? = null
                        if (imageFile != null)
                            actualImageFile = File(song.mFile.parent, imageFile.mFile.name)
                        if (actualImageFile != null && actualImageFile.exists())
                            filesToRefresh.add(imageFile!!)
                    }
                }
            }
        }
        return filesToRefresh
    }
}
